reps/2023-12-14-virtual-cluster.md

```diff
@@ -0,0 +1,452 @@
+
+### Should this change be within `ray` or outside?
+
+Inside `ray` project since this is a Ray Core feature.
+
+## Stewardship
+
+### Required Reviewers
+
+@ericl, @stephanie-wang, @scv119
+
+### Shepherd of the Proposal
+
+@ericl
+
+## Design
+
+With the introduction of virtual clusters, every Ray job runs in its own virtual cluster and only has access to (logical) resources inside that virtual cluster. Each virtual cluster has a spec that defines the min and max resources of the cluster. Min resources are minimal resources required for the job to run and they are atomically reserved for gang scheduling. If min resources cannot be reserved when there are not enough available resources, the job will be queued. With job queueing, we can implement different policies such as FIFO or priority-based queueing. Max resources are the autoscaling limit of the virtual cluster and the maximal resources can be used by the job.
+
+Virtual clusters can be nested and a Ray job can create sub-clusters to isolate separate parts of its application workload. For example, a Tune grid sweep job can create a sub-cluster for each of its nested Train workload. These possibly nested virtual clusters form a tree where the root is the entire Ray cluster.
+
+<img src="https://user-images.githubusercontent.com/898023/291139618-0be11470-db09-466d-8c2c-37b9e5b3765c.png" alt="drawing" width="508"/>
+
+Virtual clusters with different min and max resources are autoscalable. When scaling up, virtual clusters will try to borrow more resources from their parent clusters. If their parents have available resources, the scaling up is instant. Otherwise, parents will try to borrow resources from their parents recursively and eventually this may cause the upscale of the physical cluster. When scaling down, virtual clusters will return resources to their parents recursively and eventually this may cause the downscale of the physical cluster.
+
+Virtual nodes of a virtual cluster can be either fixed size or flexible/resizable. For a single virtual cluster, there can be multiple fixed-size virtual nodes but at most one flexible virtual node on a single physical node. The upscaling of a virtual cluster can be achieved by adding new virtual nodes or scaling up an existing flexible virtual node.
+
+### API
+
+#### Virtual Cluster Spec
+
+```
+message VirtualCluster {
+  // A virtual cluster consits of flexible resources and fixed size resources.
```

Comment:
Should add `map<string, string> labels` that applies to both flex and fixed nodes.

Reply:
What's the use case for adding labels to flex nodes?

Reply:
You need labels in order to schedule tasks within an specific virtual cluster in the first place, right? This is also for consistency with the physical node interface.

Reply:
Yea, flex nodes also have system labels (e.g. vc_id). I'm just not sure whether we need to expose a way for users to set custom labels for flex nodes. 

So you are suggesting we have `map<string, string> cluster_level_custom_labels` that applies to all virtual nodes of a virtual cluster. And later on fixed nodes can also set additional custom labels individually?

Or we have `map<string, string> flex_nodes_custom_labels` that only applies to flex nodes.

Reply:
I see, if we have automatically added labels for vc_id then I agree this is more of a P1.